### PR TITLE
feat(*) add key helm.sh/hook-delete-policy to hook annotation

### LINKED ISSUE #1769: Helm should delete job after it is successfully finished
I have a job bind to `post-install,post-upgrade,post-rollback`
```  annotations:
    "helm.sh/hook": post-install,post-upgrade,post-rollback
```
when i’m running update charts i get `Error: UPGRADE FAILED: jobs.batch "opining-quetzal-dns-upsert" already exists`
`kubectl get jobs` returns
```NAME                         DESIRED   SUCCESSFUL   AGE
opining-quetzal-dns-upsert   1         1            12m
```
So, how we are supposed to use jobs as hooks if it is not deleted after it is successfully finished?
There is no way to update chart if it has such job.

### PR DESCRIPTION
**Pull request by DoctorZK**
When "helm.sh/hook-delete-policy: hook-succeed" is provided in a hook's annotation, Tiller will automatically delete the hook after the hook is succeed. When "helm.sh/hook-delete-policy: hook-failed" is provided in a hook's annotation, Tiller will automatically delete the hook after the hook is failed.

Closes #1769

### CODE REVIEW

**Comment by thomastaylor312** [comment]
I am putting this in 2.7 for now because 2.6 is focused on stability. If that is going to be a big problem, we are open to discuss it 🙂

**Comment by thomastaylor312** [comment]
@DoctorZK Sorry for the delay on this. I'll try to get a review done later in the week and then we can merge after 2.6 gets cut

**Comment by DoctorZK** [comment]
OK. Please feel free to contact me if you have any question about the pr.

**Review by thomastaylor312 — CHANGES_REQUESTED** [review]
One small naming change needed. Could you also add some documentation for this?

Otherwise, the functionality is all tested and good to go. I am super excited to see this feature!

File: pkg/tiller/hooks.go [hunk 1]
```diff
@@ -173,6 +187,22 @@ func (file *manifestFile) sort(result *result) error {
...
 
 		result.hooks = append(result.hooks, h)
+
+		isKnownDeletePolices := false
+		dps, ok := entry.Metadata.Annotations[hooks.HookDeleteAnno]
+		if ok {
+			for _, dp := range strings.Split(dps, ",") {
+				dp = strings.ToLower(strings.TrimSpace(dp))
+				p, exist := deletePolices[dp]
+				if exist {
```
Comment (DoctorZK) [inline pkg/tiller/hooks.go]: OK， I will handle this bug in a new PR

File: pkg/hooks/hooks.go [hunk 2]
```diff
@@ -40,6 +43,12 @@ const (
 	ReleaseTestFailure = "test-failure"
 )
 
+// Type of policy for deleting the hook
+const (
+	HookSucceed = "hook-succeed"
```
Comment (DoctorZK) [inline pkg/hooks/hooks.go]: Excellent suggestion. I will fix this soon

File: pkg/tiller/hooks.go [hunk 3]
```diff
@@ -113,6 +119,13 @@ func sortManifests(files map[string]string, apis chartutil.VersionSet, sort Sort
 //		annotations:
 //			helm.sh/hook: pre-install
 //
+// To determine the policy to delete the hook, it looks for a YAML structure like this:
+//
+// kind: SomeKind
+// apiVersion: v1
+// metadata:
+// 		annotations:
```
Comment (DoctorZK) [inline pkg/tiller/hooks.go]: Good Suggestion

**Comment by thomastaylor312** [comment]
Thank you for the fixes @DoctorZK! Can you add some documentation for the new feature? It should be good to go once we have that

**Comment by DoctorZK** [comment]
Sure

**Comment by thomastaylor312** [comment]
Thank you for the docs. This is good to go!

**Review by thomastaylor312 — APPROVED** [review]

**Comment by redbaron** [comment]
How can I achieve that Job is deleted either way? Successful or not, I don't want it to block further updates of a given release.

**Comment by DoctorZK** [comment]
It would not block further updates since hook deletion is an asynchronous request. Also because of the asynchronous request, we can not achieve that the Job is deleted or not. What we do now is just sending the request to K8S apiserver and making sure the apiserver hase successfully recieved the request.

**Comment by redbaron** [comment]
@DoctorZK, maybe I misunderstood code, but it seems that async request to delete Job will be sent if Job succeeded or fail, there is no way to specify "delete job irrespective of exit status", right?  So there can be a situation where subsequent helm release will fail due to job still present and there is no way to prepare helm chart in such way that it works for all the cases, in other ways I can't say:

```
"helm.sh/hook-delete-policy": ["hook-succeeded", "hook-failed"]
```

is this case covered somehow else?

**Comment by thomastaylor312** [comment]
@redbaron You can specify both. I tried it and it deleted it irrespective of the exit code

**Comment by thomastaylor312** [comment]
Oh, but the syntax is `"helm.sh/hook-delete-policy": "hook-succeeded,hook-failed"`

**Pull request merged**